### PR TITLE
CDRIVER-2168: Escape JSON strings for regex, symbol, and DBPointer types

### PR DESCRIPTION
**Pull request by jmikola**
https://jira.mongodb.org/browse/CDRIVER-2168

### CODE REVIEW

File: src/bson/bson.c [hunk 1]
```diff
@@ -2792,9 +2801,16 @@ _bson_as_extended_json_visit_dbpointer (const bson_iter_t *iter,
 {
    bson_json_state_t *state = data;
    char str[25];
+   char *collection_escaped;
+
+   collection_escaped = bson_utf8_escape_for_json (v_collection, v_collection_len);
```
Comment (jmikola) [inline src/bson/bson.c]: Per [collection naming restrictions](https://docs.mongodb.com/manual/reference/limits/#naming-restrictions), it's possible for a collection name to contain special characters that require escaping.

File: src/bson/bson.c [hunk 2]
```diff
@@ -2728,9 +2728,16 @@ _bson_as_json_visit_regex (const bson_iter_t *iter,
 {
    bson_json_state_t *state = data;
    const char *c;
+   char *escaped_regex;
+
+   escaped_regex = bson_utf8_escape_for_json (v_regex, -1);
```
Comment (jmikola) [inline src/bson/bson.c]: As mentioned in my comment in [CDRIVER-2168](https://jira.mongodb.org/browse/CDRIVER-2168), this causes the [regex with slash](https://github.com/mongodb/specifications/blob/0fc09a2bc56de5fbee6f9c046083d39b5b62faf2/source/bson-corpus/tests/regex.json#L22) test from the BSON corpus suite to fail.

@xdg cited the [section 7 of RFC-7159](https://tools.ietf.org/html/rfc7159#section-7) in a related HELP ticket as evidence that `bson_utf8_escape_for_json()` need not do this. Additionally, [JSON.org](http://www.json.org/) suggests that escaping of forward slashes is optional.

Comment (bjori) [inline src/bson/bson.c]: Just to clarify. The escaping done in `bson_utf8_escape_for_json` includes escaping of "optional characters".

The bson corpus tests suite has a test that expects the optional characters *not* to be escaped.

I read this as both libbson and the bson corpus are both correct.

@xdg How do you suggest we handle it? Duplicate the corpus test and modify it?
Add a test case to the corpus, and mark then mutually exclusive?

Comment (xdg) [inline src/bson/bson.c]: As with whitespace and key order, tests need to account for legal differences between the corpus text and the generated text.  For example, in Perl, I can "wash" the JSON texts through a decode->encode cycle that normalizes differences.  For C, I imagine that tests will need to iterate keys and compare values directly in some way that accounts for character escaping differences while still ensuring that all *required* characters are escaped.

**Comment by jmikola** [comment]
Closing, as @ajdavis implemented his own patch.

**Pull request closed without merging**